src/plugins/custom/authorization.ts

```diff
@@ -0,0 +1,39 @@
+import fp from 'fastify-plugin'
+import { FastifyReply, FastifyRequest } from 'fastify'
+
+declare module 'fastify' {
+  export interface FastifyInstance {
+    isModerator: typeof isModerator;
+    isAdmin: typeof isAdmin;
+  }
+}
+
+function verifyAccess (request: FastifyRequest, reply: FastifyReply, role: string) {
+  if (!request.session.user.roles.includes(role)) {
+    reply.status(403).send('You are not authorized to access this resource.')
+  }
+}
+
+async function isModerator (request: FastifyRequest, reply: FastifyReply) {
+  verifyAccess(request, reply, 'moderator')
+}
+
+async function isAdmin (request: FastifyRequest, reply: FastifyReply) {
+  verifyAccess(request, reply, 'admin')
+}
+
+/**
+ * The use of fastify-plugin is required to be able
+ * to export the decorators to the outer scope
+ *
+ * @see {@link https://github.com/fastify/fastify-plugin}
+ */
+export default fp(
+  async function (fastify) {
+    fastify.decorate('isModerator', isModerator)
+    fastify.decorate('isAdmin', isAdmin)
```

Comment:
why arent' we doing `request.isAdmin()` instead?

Reply:
Hum, now that I am thinking about it.
I am not sure.